lib/active_model/serializer/concerns/associations.rb

```diff
@@ -85,18 +85,26 @@ def associate(reflection)
       #
       def associations(include_directive = ActiveModelSerializers.default_include_directive, include_slice = nil)
         include_slice ||= include_directive
-        return unless object
+        return Enumerator.new unless object
 
         Enumerator.new do |y|
-          self.class._reflections.values.each do |reflection|
+          self.class._reflections.each do |key, reflection|
             next if reflection.excluded?(self)
-            key = reflection.options.fetch(:key, reflection.name)
             next unless include_directive.key?(key)
 
-            y.yield reflection.build_association(self, instance_options, include_slice)
+            y.yield reflect_association(reflection, key: key, include_slice: include_slice)
           end
         end
       end
+
+      # @param reflection [Reflection]
+      # @param association_options [Hash] required: include_slice
+      # @return [Association]
+      def reflect_association(reflection, association_options)
```

Comment:
I agree with preferring composition. I'd love to get there.  

I believe the mixins were originally extracted to make the serializer easier to understand, but ultimately just made the complexity harder to see.

My comment about the `_` was specifically about going out of our way to define instance methods on the serializer that do not conflict with user-defined methods.